Source/UIKit/DataSource.swift

```diff
@@ -160,52 +160,74 @@ open class DataSource<Item: Equatable, View: DataSourceView> : NSObject {
 
 extension DataSource {
     
-    func underlyingDataChanged(_ new: [Item], _ change: ArrayChange<Item>) {
-        items = Array(new)
-        
-        if suppressChangeUpdates == false {
-            // Need to use straight up `reloadData()` if the table view / collection view has not done the initial data load yet
-            // (for example, if the collection view is on a page that is not yet visible).
-            //
-            // Reason being, when we do the insertCells / deleteCells calls (and the view hasn't loaded any data yet), it'll do a query
-            // "before" to get the old ("current") counts, but we've already updated our internal items array, so we return the new items.
-            // Therefore, when we try to do the insertCells / deleteCells, we get exceptions thrown like:
-            //
-            //      *** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'attempt to delete 
-            //      item 3 from section 0 which only contains 1 items before the update'
-            //
-            if animateChanges && hasDoneInitialDataLoad {
-                switch change {
-                case .set(elements: _):
-                    view?.reloadData()
-                    
-                case let .insert(index: index, newElements: newElements):
-                    let indexPaths = (index ..< index + newElements.count).map { i in IndexPath(item: i, section: 0) }
-                    view?.insertCells(indexPaths: indexPaths)
-                
-                case let .remove(range: range, removedElements: _):
-                    let indexPaths = range.map { i in IndexPath(item: i, section: 0) }
-                    view?.deleteCells(indexPaths: indexPaths)
-                    
-                case let .replace(range: range, removedElements: _, newElements: new):
-                    view?.batchUpdates { [view = view] in
-                        let deleted = range.map { i in IndexPath(item: i, section: 0) }
-                        view?.deleteCells(indexPaths: deleted)
-                        
-                        let addedRange = range.lowerBound ..< range.lowerBound + new.count
-                        let added = addedRange.map { i in IndexPath(item: i, section: 0) }
-                        view?.insertCells(indexPaths: added)
-                    }
-                }
-            }
-            else {
-                view?.reloadData()
+    func underlyingDataChanged(_ new: [Item]) {
+        let change = items.calculateChange(new)
+
+        guard suppressChangeUpdates == false else {
+            items = new
+            syncSelections()
+            return
+        }
+
+        guard animateChanges else {
+            items = new
+            view?.reloadData()
+            syncSelections()
+            return
+        }
+
+        if case .set = change {
```

Comment:
I think you only need the `if case` syntax if you want to access an associated value though, IIRC. So I _think_ `&& change == .set` would work…